### ISSUE TITLE
Communicating with a Hella actuator over CAN

### ISSUE THREAD

**Issue by MatLAdams** [djwlindenaar/Hella_Turbo_Controller#1]
Hi, ive been reading your Git hub page, im completely6 new to Github so undersatnd very little about how it works, ive read that the only way to communicate with you is to open an issue, so here i am.

Im intending to use a Hella turbo actuator to control a gearbox on a car. I initially tried the most basic version of the Hella actuator with no on-board crontroler, just termainals on to the notor and a PWM position feedback for the seperate motor controller to know where the actuators at. This version quickly broke when i was trying to use DC on the motor - i think the gears were also quite worn as it was an old actiator.

Im now trying a PWM / CAN actuator 6NW008412. Im using a Korlan USB2CAN interface with Ubunti and the CAN_UTILS suite. Running cansniffer i can see feedback messages coming in on ID 4EB. The fourth pair of digits appear to report position  - i can move the actuarot arm a small amount due to a worn gearbox and see the position updating between 64 and88. The last pair of digits apear to report motor activity.

I would really appreciate any help controlling the actuator position with CAN, i read your Wiki but it was a little difficult for me to understand. I tried sending a couple of messages using CANSEND but saw no movement from the actuator.

Thanks, in advance, Mat

**Comment by djwlindenaar** [comment]
Hi Mat, 
nice idea! 
I think I can help, but I have to admit that this project has lost my attention somewhat.  I find your controller quite interesting, since the only ones I've seen were configured to be controlled through PWM input. Therefore I'd be interested to see the configuration values in the eeprom.

I'll have to dig up the project and check how I can get you in a position to read the eeprom. That should also help to identify the CAN message used to control the position. Or if you like, you can configure it to be controlled by PWM to set the target position.

Best regards
Daniel

**Comment by djwlindenaar** [comment]
@MatLAdams , I've committed a python script in this repository, which you can try to execute on your system. 

Near the bottom, you can see a class being instantiated with parameters, which should match your setup. Please fill the correct canbus identifier (same as you use with cansend).

After executing it, there should be a `DATE-TIME.bin` file in the same folder. Please run that through xxd and paste in your reply. I'll get back to you based on that information.

Best regards
Daniel

**Comment by MatLAdams** [comment]
Hi Daniel,

Thankyou for this. I will take a good look at it over the next day so so and see what I can figure out. I’m new to Linux/Ubuntu as I only just started using it for a different CAN project. Python I have no idea about  at all but I’ll see where I get. The one thing I did see in the script were referenced to SLCAN and SOCKETCAN. My can interface (Korlan USB2CAN is addressed within Linux simply as Can0 as it’s not a serial can or virtual can device. Unfortunately I couldn’t get socket can to work when I was trying to use CAN software called Kayak and don’t know from looking at your code if that’s going to be a problem this time.

Anyway, thanks again and I’ll see what happens.

Best wishes, Mat

**Comment by MatLAdams** [comment]
Hi Daniel,

I'm back on this now after success with using a different actuator with PWM. I have sufficient control with this interface for what i need however i have no feedback to my controller which isn't ideal.

Im not having much success running your script as i don't understand python. I found the class being defined but this was at the top of the script, is this what you meant - its the only reference to 'class'. I amended the lines i found from slcan0 to can0; when running the script it returned :-

parallels@ubuntu:~$ python mod_hella_prog.py
Traceback (most recent call last):
  File "mod_hella_prog.py", line 3, in <module>
    import can
ImportError: No module named can
parallels@ubuntu:~$ ^C

Clearly, i don't know what i'm doing. If you have the time and inclination to take me through it step by step i'd be more than happy to oblige but if not i might have to bow out at this point.

Thanks for your help thus far, best wishes Mat.

**Comment by djwlindenaar** [comment]
Sorry @MatLAdams for being silent for so long. I must have missed your post.

Anyway. The issue is that your python install doesn't have the python-can class installed. You can get that by running `pip3 install python-can`.

If pip3 doesn't work, you should use your regular package manager to install python3-pip...

Good luck!

**Comment by test3210-d** [comment]
@djwlindenaar
If you interesting i can give you g-221 eprom
diffegent motor directin, and not work 3f0 id  (config id eprom 0x29=0x2a )

**Comment by djwlindenaar** [comment]
@test3210-d 
Certainly interested! Where did you get the controller from? What application, car?

**Comment by test3210-d** [comment]
[G-221.zip](https://github.com/djwlindenaar/Hella_Turbo_Controller/files/5119951/G-221.zip)
Ford Mondeo 2.0 TDCi 6NW008412, G-221 i write eprom only, 
config address in eprom address 41 , bit 4 -get pwm from can , bit 6 enable can tx 0x4eb , request id 0x04ea 
in some actuators (g-276 ) enabled can tx and pwm input config 0x20  tx id 0000 (error ?? eeprom from friends database) 
where can id config i don't know now, but work in progress deniss979@gmail.com

**Comment by test3210-d** [comment]
@djwlindenaar i find response can id address 39
value x8  (0x22*8=0x110 , 0x9d*8=0x4eb)

**Comment by test3210-d** [comment]
i try different eepron decode and now i know
stored 2 value for can control, can response config 39  (byte 1) and 40 (byte2) address  and request id 36 and 37 address.
Data format byte1 *8 + byte 2>>5 ; packet data size byte 2 & 0xf 
May be add to wiki ?
bit 1 eeprom 0 - motor direction 
can command id have same structure address 9 and 10 but tesponse id = command id -8 

**DANGER** 
I brick 2 module by write incorrect value into control registers one in 41 fmd one in 10

**Comment by djwlindenaar** [comment]
@test3210-d , well it looks like I missed your further comments some time ago! sorry about that.

How do you feel about contributing to the wiki?

**Comment by peterschumann** [comment]
@djwlindenaar I tried to get CAN comms working on actuator out of a Mercedes OM648. They are PWM controlled in the original application but also have the CAN Hardware inside. 
But it doesn't broadcast any CAN Frames and I see no reactions if I send a frame with ID 0x3F0.
Is there a way to get CAN working? I assume its disabled in the eeprom...

**Comment by djwlindenaar** [comment]
@peterschumann , well, I'm not sure. Maybe it's possible that the can interface can be disabled programmatically. Although that seems strange since it's used while adjusting the actuator to the turbo. 

I'd double-check the CAN interface from an electronics p.o.v.. Right pins connected? Right bit rate? Does the controller actually send the commands? Anything else on the bus disrupting communication?

Possibly there's a way to change the CAN ID for requests (Instead of 0x3F0). You could try to send a message to all possible IDs and see if you get any response at all.

In short: I'm not sure 😊

Possibly you'll have to find another donor

**Comment by test3210-d** [comment]
All function can be changed in eprom . Control and programming CAN Id, pwm freq , start-stop angle, motor rotation (cw or ccw) , enable - disable can commands, serial number , etc... Give me eprom and i say all data
if you know G-xxx code i can search data in own database
you can try id 
0x3f0
0x61c
0x618
0x660

**Comment by peterschumann** [comment]
@djwlindenaar I used a scope to check the CAN Bus Frames, the scope does CAN decoding and the frames looked good. But there was no CAN ACK, which is an indication for no CAN Node is active. The ACK should be there in any case, even if the ID is wrong. 
Maybe there is something like a boot mode for the initial configuration, that maybe also useful for recovery of your bricked ones @test3210-d  ?

**Comment by peterschumann** [comment]
The two I tried are G-109, both work fine with PWM. But nothing over CAN. 
I will try the CAN IDs next week

**Comment by djwlindenaar** [comment]
No ack is indeed an issue... Polarity correct?

If you can find a kind of "factory reset" that would be great. Ability to recover from bricks is always a reassuring thing while hacking. :)

**Comment by test3210-d** [comment]
G-109 it is pwm controlled 140hz witch 0x3f0 programming id and don't have can bus control id
try send CAN request for ID 0x3F0   data to send 0x49,0x00,0x00,0x00,0x00,0x00,0x00,0x00  response must be 
data length =8 !!!

**Comment by s-rv-s** [comment]
Hi @test3210-d, can you help me with connect to actuator via CAN?... 
its marked as hella 6nw008091, also have another label -  alliedsignal 704063
so i was try send CAN request for ID 0x3F0 (as actual value .py code in repo)  but newer got any answer from controller.
actuator on the car controlled via PWM 300Hz notmally.
G code can be G105 and G106 or G209 and G210 (i have 2 actuators on my engine)
or maybe my actuators it just not supports to programming via CAN.....

thanks for answer

**Comment by Sergio2048** [comment]
Hi everyone, can i use 6nw008412 from. Ford TDCI with only pwm pin ? Is it 140hz or 300hz? Thank you

**Comment by awpe** [comment]
Seems like different actuators have very different IDs they use. From ECU firmware I was able to extract status IDs and control IDs for my G-22 actuator. However it uses 4 bytes long data frames to report status/position and command new position. Also position has to be repeated each 500ms otherwise it goes to 0. All IDs from wiki are ignored on that one.

@test3210-d Could you please share with control IDs for G-22 actuator? Status one is 0x71C, and set position one is 0x386, but no IDs from that wiki is accepted.

**Comment by dirtydevious** [comment]
Has this project died. It doesn't seem to have had any reply or update from the originator

**Comment by dirtydevious** [comment]
> The two I tried are G-109, both work fine with PWM. But nothing over CAN. I will try the CAN IDs next week

Did you have any luck with the CAN IDs?

**Comment by dirtydevious** [comment]
This is totally possible and would be great if we could come up with a DIY open source version of this: https://turbotechnics.com/atp-100/ or this https://turbo.it/en/etp-tester-and-programmer/  https://youtu.be/o_vvnAsGt8E

**Comment by jlaustill** [comment]
I'm looking for a purely canbus controlled actuator. Does anyone know a part number for one? It looks like the G-22 might be promising, but I can't find one anywhere. @awpe does yours have any other numbers on it?

**Comment by awpe** [comment]
> I'm looking for a purely canbus controlled actuator. Does anyone know a part number for one? It looks like the G-22 might be promising, but I can't find one anywhere. @awpe does yours have any other numbers on it?

Hello, G22 is a gearbox number (direction, ratio). Can bus have REA connectors and pwm ones have SREA.

**Comment by jlaustill** [comment]
@awpe do you mean that the numbers for the gearbox have meaning? So 22 is direction 2? like 2 oclock maybe? and 2 is ratio? like 2:1 or something? I found a G-022 that's for a VW PHAETON (3D1, 3D2, 3D3, 3D4, 3D6, 3D7, 3D8, 3D9) 2002-2006 with a 5.0 V10. It has these reference numbers listed 
```
755297-4
755297-1
755297-0001
755297-3
755297-0003
755297-4
755297-0004
755297-5
755297-0005
742805-1
742805-0001
742805-3
742805-0003
742805-5
742805-0005
07Z145907G
07Z145907GV
07Z145907GX
07Z145873G
07Z145873E
07Z145701R
07Z145701P
```

The arm is pointing in the perfect direction for my use case, so if it's canbus and works with `control IDs for G-22 actuator? Status one is 0x71C, and set position one is 0x386, but no IDs from that wiki is accepted.` then I am golden, I can figure out which bytes do what by just playing with it!

**Comment by jlaustill** [comment]
@awpe if my research is proving correct, then the G-022 used on the VW Phaeton would be pin 1 12v, pin 2 ground, pin 3 CANH, and pin 5 CANL, does that sound right?

**Comment by Baryonn** [comment]
@test3210-d  Hi, would you be open for a couple of questions regarding the reading and writing of eprom of Hella actuator?

**Comment by jlaustill** [comment]
Ok, I got a G-222 actuator. It's a legit Hella made in Germany with the numbers 712120 and 6nw 008412. I'm getting status messages on 0x658 that look like this "ID: 658 Buffer: 4 0 0 D6 0 1D 0 0" I want to figure out which id I need to send to control the thing. Here's the output from the memory 
```jlaustill@jlaustill-Precision-7730:~/code/Hella_Turbo_Controller$ xxd test.bin
00000000: 3635 3a20 3038 0a36 363a 2030 300a 3637  65: 08.66: 00.67
00000010: 3a20 4342 0a36 383a 2030 380a 3639 3a20  : CB.68: 08.69: 
00000020: 3632 0a36 413a 2041 300a 3642 3a20 3936  62.6A: A0.6B: 96
00000030: 0a36 433a 2030 360a 3644 3a20 3032 0a36  .6C: 06.6D: 02.6
00000040: 453a 2031 440a 3646 3a20 3039 0a37 303a  E: 1D.6F: 09.70:
00000050: 2046 460a 3731 3a20 4646 0a37 323a 2046   FF.71: FF.72: F
00000060: 460a 3733 3a20 4646 0a37 343a 2046 460a  F.73: FF.74: FF.
00000070: 3735 3a20 4646 0a37 363a 2032 300a 3737  75: FF.76: 20.77
00000080: 3a20 4646 0a37 383a 2032 300a 3739 3a20  : FF.78: 20.79: 
00000090: 3230 0a37 413a 2030 410a 3742 3a20 3534  20.7A: 0A.7B: 54
000000a0: 0a37 433a 2030 310a 3744 3a20 3845 0a37  .7C: 01.7D: 8E.7
000000b0: 453a 2030 310a 3746 3a20 4137 0a         E: 01.7F: A7.
```
I'm not really sure what to make of the data though, can anyone point me in the right direction?

**Comment by jlaustill** [comment]
Ok, I have the G-222, 6NW 008 412 working on my turbo. I designed a bracket in cad and had it cut out by send cut send and it's working great. I have the data from the canbus decoded to get load, position, status, and temp. BUT, it is PWM controlled. I really wish I could understand how to update it to be controlled over the canbus too without bricking it. If anyone can help me with that I would be eternally grateful, and I'm offering a 100USD cash bonus...

**Comment by awpe** [comment]
I suppose the best bet is to get an actuator with can control enabled and adapt your control software to it rather than waiting for someone to disclose proprietary info on the eeprom bytes meaning. For sure you can eliminate obvious bytes from eeprom like position related ones and then experiment with some script that will change 1s and 0s in all other bytes while checking can responsiveness for position cmd, but again Id bet it will take hell of a time compared to just getting any used actuator which is same dimensions but already can driven...

**Comment by jlaustill** [comment]
The adapting the software is the easy part. Its finding an actuator with the right angle that is the hard part. The G-222 has the perfect angle for my DPS Turbonators. I just want it to be canbus controlled so I don't have to manage the dang PID for the PWM control. I know it's proprietary, but isn't that why this project exists in the first place?

**Comment by awpe** [comment]
what do you mean by angle? Angle is programmable by that min/max, then you ask for pos by sending 0-1023 request in 2 bytes (first is 0-3, second 0-255), all of those actuators are 120deg absolute with programmable angle... Diffs are direction, geometry of lever and pwm/can control. So you can grab anything with the same direction and suiting lever, then set min max using that software here and you are done

**Comment by awpe** [comment]
I was not able to truly identify the meaning of bytes as it seems like there are other eeprom addresses not covered by this software and some bits should be set in a non obvious manner though I didn't invest any serious time in experimenting...

**Comment by Alexkorn1984** [comment]
Hello. I'm trying to write down the minimum and maximum values ​​at the addresses 0x03-0x06, as well as in 0x22 . And I still have an interval smaller than the difference between them. Maybe someone knows if there is another cell somewhere where I need to write something?

**Comment by jlaustill** [comment]
Hi all, I've been doing extensive reverse engineering on both a G-222 and G-221 actuator and wanted to share what I've confirmed. Full details and code at https://github.com/jlaustill/Hella_Turbo_Controller

## CAN Position Control - Confirmed Working

I got full CAN position control working on both a G-221 (native CAN) and a G-222 (converted from PWM to CAN via EEPROM changes).

### G-221 (Ford/Jaguar 2.0 TDCi)
- **Command ID: 0x4EA** — bytes[0:1] big-endian 16-bit position, continuous ~20ms
- **Broadcast ID: 0x4EB** — bytes[2:3] big-endian 16-bit position feedback
- **Programming: TX=0x660, Response=0x658, Ready=0x65B**
- Commands must be sent continuously or the actuator returns to default (watchdog)

### G-222 Converted to CAN
The G-222 can be converted to accept CAN position commands by changing these EEPROM bytes:

| Addr | Original | New | Purpose |
|------|----------|-----|---------|
| 0x04 | unit-specific | physical min endstop | Min position |
| 0x06 | unit-specific | physical max endstop | Max position |
| 0x22 | 0x00 | 0xD0 (unit-specific) | Rotation offset — calibrates position sensor |
| 0x24-0x25 | 0x00 0x08 | 0x9D 0x48 | Command CAN ID = 0x4EA |
| 0x27-0x28 | 0xCB 0x08 | 0x9D 0x68 | Broadcast CAN ID = 0x4EB |
| 0x29 | 0x62 | 0x2A | Mode register — enables CAN position control |

All changes must also be written to mirror addresses (+0x40). Power cycle required after changing 0x29.

After conversion, command format on 0x4EA is **single byte** in byte[0]:
- 0-250 = linear position (0=min, 250=max)
- 251-255 = return to default position
- Position feedback on 0x4EB bytes[2:3] = command * 4

### Key Findings

**Mode register (0x29) is an enum, not a bitmask.** Firmware validates values:
- 0x62 = PWM mode (G-222 default)
- 0x2A = CAN position control mode (G-221 default)
- 0x72 and 0x6A are rejected

**CAN IDs are stored in EEPROM** with encoding: `high_byte * 8 + (low_byte >> 5)`
- 0x09-0x0A = Programming response ID
- 0x24-0x25 = Position command ID
- 0x27-0x28 = Position broadcast ID

**Rotation offset (0x22)** is unit-specific — depends on physical sensor alignment. Setting it wrong causes position feedback to read 1001 or 0. Find the right value by sweeping and checking when cmd=0 reports pos=0.

**Min/max (0x03-0x06)** should be set to actual physical endstops found using register-level motor drive (mode=5, register 0x63).

This confirms what @test3210-d found earlier — address 41 (0x29) controls the mode, and addresses 36-37 (0x24-0x25) and 39-40 (0x27-0x28) are the CAN IDs. The "bit 4" observation was based on 1-based bit numbering.

### PCB Notes
- G-221 and G-222 have **identical PCBs** (both use L9805E chip)
- G-277 has a completely different PCB — no CAN traffic at any bitrate
- Gearbox/sensor assemblies are NOT interchangeable between units (different sensor alignment)

**Comment by tboHammu** [comment]
@jlaustill I want to give a huge thanks for your findings & development, I was sold a 6NW 009 228 G-22 servo as a PWM controlled one when in fact it is CAN only (Touareg/Phaeton unit most likely). After fiddling with the CAN ID's I was able to read the EEPROM off of it and indeed, writing value 0x62 to addresses 0x29 and 0x69 converted it into 300Hz PWM control!

Attached is the backup from the mentioned servo

[20260707_220339_memory_dump.zip](https://github.com/user-attachments/files/29765017/20260707_220339_memory_dump.zip)

**Comment by marccaro** [comment]
Hello, I found this project and I want to attemp to use this to make my edc17cp20 use a hella actuator. For my interest I need a pwn one but i need to be able to read the position via can and send back to the ecu via 0-5v analog signal. If the actuator is in pwn mode the position is still broadcasted? Is it fast enough to use as a feedback for closed boost loop control? If anyone with experience could tell me what 6NW XXXXXXXXX would suite me more would be really helpfull
